### PR TITLE
Adding better redis error handling

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: server/data/redisClient.ts [hunk 1]
```diff
@@ -9,10 +10,22 @@ const url =
...
+      reconnectStrategy: (attempts: number) => {
+        // Exponential back off: 20ms, 40ms, 80ms..., capped to retry every 30 seconds
+        const nextDelay = Math.min(2 ** attempts * 20, 30000)
+        logger.info(`Retry Redis connection attempt: ${attempts}, next attempt in: ${nextDelay}ms`)
+        return nextDelay
+      },
+    },
   })
+
+  client.on('error', (e: Error) => logger.error('Redis client error', e))
```
Comment (andrewrlee) [inline server/data/redisClient.ts:28]: needed else unhandled exceptions will bring down server

File: server/middleware/authorisationMiddleware.ts [hunk 2]
```diff
@@ -18,5 +19,5 @@ export default function authorisationMiddleware(authorisedRoles: string[] = []):
 
     req.session.returnTo = req.originalUrl
```
Comment (andrewrlee) [inline server/middleware/authorisationMiddleware.ts:20]: setting things on the session may fail due to redis network errors

**Review by psoleckimoj — APPROVED** [review]
LGTM!

File: server/data/redisClient.ts [hunk 3]
```diff
@@ -9,10 +10,22 @@ const url =
...
     password: config.redis.password,
     legacyMode,
+    socket: {
+      reconnectStrategy: (attempts: number) => {
+        // Exponential back off: 20ms, 40ms, 80ms..., capped to retry every 30 seconds
+        const nextDelay = Math.min(2 ** attempts * 20, 30000)
+        logger.info(`Retry Redis connection attempt: ${attempts}, next attempt in: ${nextDelay}ms`)
+        return nextDelay
+      },
+    },
```
Comment (andrewrlee) [inline server/data/redisClient.ts:25]: The default reconnect strategy in v4 continually retries to connect which maxes out CPU [bug](https://github.com/redis/node-redis/issues/2095)

File: server/index.ts [hunk 4]
```diff
@@ -8,7 +8,7 @@ import UserService from './services/userService'
 
 promClient.collectDefaultMetrics()
 
-const hmppsAuthClient = new HmppsAuthClient(new TokenStore(createRedisClient()))
+const hmppsAuthClient = new HmppsAuthClient(new TokenStore(createRedisClient({ legacyMode: false })))
```
Comment (petergphillips) [inline server/index.ts:11]: should this be `true` instead?

Comment (andrewrlee) [reply server/index.ts:11]: want this to be false so can use async api

File: server/data/redisClient.ts [hunk 5]
```diff
@@ -9,10 +10,22 @@ const url =
     ? `rediss://${config.redis.host}:${config.redis.port}`
     : `redis://${config.redis.host}:${config.redis.port}`
 
-export const createRedisClient = (legacyMode = false): RedisClient => {
-  return createClient({
+export const createRedisClient = ({ legacyMode }: { legacyMode: boolean }): RedisClient => {
```
Comment (psoleckimoj) [inline server/data/redisClient.ts:13]: This blew up on our app as there is no `return` any more?

**Pull request merged**